Added conformers interface to plams

GiulioIlBen opened the pull request:
The plams interface instead of living in `scm.conformers` has moved to `scm.plams`. This is in accordance with what @mhellstr and I discussed, we think is a cleaner approach and more inline with plams design.

Added method `plot_conformers` to `ConformersResults` .

Updated the notebooks examples but not the .py files yet. 

Let me know what do you think.

Commit:
Added conformers interface to plams

Commit:
black formatting

dormrod commented on examples/COSMORSConformers/cosmors_conformers.ipynb:
```diff
@@ -18,14 +18,14 @@
     "from scm.plams.recipes.adfcosmorsconformers import ADFCOSMORSConfJob, ADFCOSMORSConfFilter\n",
     "from scm.plams import Molecule, from_smiles, Settings, init, config, JobRunner\n",
-    "from scm.conformers import ConformersJob\n",
+    "from scm.plams import ConformersJob\n",
```
You need to run `generate_example.sh` on all these examples to get these updates into the docs

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,280 @@
+from scm.plams.tools.plot import plot_molecule
+from scm.plams.tools.units import Units
+
+from .ams import AMSJob, AMSResults
```
(convert to absolute import)

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,280 @@
+        self.molecules = list(sorted(molecules, key=lambda m: m.properties["Energy in kcal/mol"], reverse=False))
+
+    @property
+    def energies(self):
```
Please could you add any missing return type hints (including `-> None`) for all these methods whilst we are doing this refactor?

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,280 @@
+        elif isinstance(molecule, Molecule):
+            self.molecule = molecule.copy()
+        else:
+            raise NotImplementedError("TODO: Implement multiple molecules input.")
```
Please update this to something more like - `ValueError(f"If specified, molecule but be a single Molecule instance, but was '{molecule}'"`

Also add a comment about UCS... this probably needs to support it in future

Commit:
generate_example.sh updated some commands to uv

switched to uv some commands such that you do not need to pollute the amspython with tools like black and nbconvert

Commit:
run generate_example.sh

as mentioned in https://github.com/SCM-NV/PLAMS/pull/319#discussion_r2348356620

Commit:
converted to absolute imports

Commit:
added type hints and better errors

https://github.com/SCM-NV/PLAMS/pull/319#discussion_r2348364948
and
https://github.com/SCM-NV/PLAMS/pull/319#discussion_r2348378743

dormrod commented on doc/source/examples/COSMORSConformers/cosmors_conformers.ipynb.rst:
```diff
@@ -8,14 +8,14 @@ Initial imports
 ::
 
-   PLAMS working folder: /path/plams/examples/COSMORSConformers/plams_workdir
+   PLAMS working folder: /home/benedini/PLAMS/examples/COSMORSConformers/plams_workdir.002
```
The generate_example should have sanitised these strings - I will have to look why this did not work

dormrod replied:
Fixed

dormrod commented on examples/ConformersGeneration/ConformersGeneration.py:
```diff
@@ -1,55 +0,0 @@
-#!/usr/bin/env amspython
-from scm.plams import *
-from scm.conformers import ConformersJob, ConformersResults
```
This is used as an example in the nightly tests - so either that needs to be updated or this can be added back in

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,315 @@
+        return el
+
+    @requires_optional_package("matplotlib")
+    def plot_conformers(
```
Should this be here or in tools/plot for consistency?

dormrod commented on CONTRIBUTING.md:
```diff
@@ -22,6 +22,7 @@ top of things.
 * Make commits of logical units
 * Make sure your commit messages are informative
 * Make sure you have added the necessary tests for your changes
+* Add examples: take a look at the corresponding [README.md](./examples/README.md) 
```
Also update CHANGELOG (here and for your change!)

GiulioIlBen replied:
should I mention that I changed the CONTRIBUTING.md file in the CHANGELOG?

dormrod committed:
Fix path sanitisation in generate_examples script

mhellstr committed:
remove awk dependence in generate_example.sh

dormrod commented on src/scm/plams/__init__.py:
```diff
@@ -52,14 +52,18 @@
     forcefield_params_from_kf,
 )
-from scm.plams.interfaces.adfsuite.inputparser import get_system_blocks_as_molecules_from_input, input_to_settings
+from scm.plams.interfaces.adfsuite.inputparser import (
```
For future PRs, please can you turn off the auto-reordering of imports? It bloats the diff and makes it a bit harder to review

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,315 @@
+__all__ = ["ConformersJob", "ConformersResults"]
+
+
+class ConformersCollection:
```
I see that this is actually not just a straight "lift and shift".

You have replaced the old `Conformers` usage to remove the dependency on `scm.conformers` with this class. This seems like it might have consequences - will some functionality be lost? The plan is I assume to remove the old implementation in `scm.conformers`?

Might be worth asking @rebadam if you have not already discussed with her.

rebadam replied:
I am quite ok with the PLAMS wrapper around the conformers tool being very basic. Though I would prefer if we could point the user to the conformers library we already have if they want more functionality. But I don't like so much that you wrote a whole new Conformers class. It is very basic, which is good, but I am afraid people will start adding to it, which will result in a lot of double coding. I would prefer if the functionality of the ConformersContainer class would just be folded into the Results class directly. The `conformers` property of the results class could just be deprecated.

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,315 @@
+        if not path.name.endswith("conformers.xyz"):
+            raise ValueError(f"{path=} not valid")
+
+        assert path.exists()
```
(don't use assert in production code)

dormrod commented on src/scm/plams/interfaces/adfsuite/conformers.py:
```diff
@@ -0,0 +1,315 @@
+            raise ValueError(f"{path=} not valid")
+
+        assert path.exists()
+        molecules = read_all_molecules_in_xyz_file(str(path))
```
Using the xyz files for example loses any bonding information?

rebadam replied:
Yes, I agree here. It would be super simple to just read from the conformers.rkf file, which also has higher precision for the coordinates. The RKF file just has the geometries and energies in the History section.

GiulioIlBen commented:
If I run ConformersJob from a different Python environment, which Python environment does the workflow use—the default amspython or the one I launched it from?

dormrod commented:
> If I run ConformersJob from a different Python environment, which Python environment does the workflow use—the default amspython or the one I launched it from?

It will be called with:
```
"$AMSBIN"/amspython -c "from scm.utils.workflows import run_workflow; from scm.conformers import AMSConformers; run_workflow(AMSConformers)" $@
```
so amspython

GiulioIlBen committed:
Merged trunk and resolving merge conflicts

GiulioIlBen committed:
mypy fixes
